src/liballoc/vec.rs

```diff
@@ -2410,9 +2410,8 @@ impl<T> Iterator for IntoIter<T> {
                     // same pointer.
                     self.ptr = arith_offset(self.ptr as *const i8, 1) as *mut T;
 
-                    // Use a non-null pointer value
-                    // (self.ptr might be null because of wrapping)
-                    Some(ptr::read(1 as *mut T))
```

Comment:
Would it be possible for `ptr::read` to contain a `debug_assert_eq!(ptr.align_offset(align_of::<T>()), 0)` ?

Reply:
My plan was to open an E-easy issue (once https://github.com/rust-lang/rust/pull/53783 lands) to decorate all these functions with a `debug_assert!` testing non-NULL and being aligned.

I would however use `ptr as usize % mem::align_of<T>() == 0`, because your test will always fail on miri. Remember, `align_offset` can spuriously fail.

Reply:
That's a good idea. 

> Remember, align_offset can spuriously fail.

I didn't know this :/ whyyyy

Reply:
Because we cannot implement it completely in miri/CTFE.

Reply:
Ah, that makes sense. This is getting off topic, but couldn't miri track the alignment requested for an allocation and use that in `align_offset` ?

Reply:
It could, and it could give a better return value in that case.

But `align_offset` is also used e.g. on a large `u8` allocation to find a subslice that is 256bit-aligned for use with SSE. There is no way for miri to provide an alignment stronger than what the allocation promises.

So maybe it could be made good enough for your test, not sure... the `%` works today though.

Reply:
It is a bit worrying that we might prefer slower run-time code because it works on miri to fast run-time code. Does the `%` lower to the same LLVM-IR than `align_offset`, at least after running `opt` ? 

An alternative is to use conditional compilation, and use `%` for `miri` and `align_offset` for run-time rust.

Reply:
I don't see how `%` could be any slower than the intrinsic. If anything, `align_offset` is doing the harder job (computing which offset to add, not just testing it for 0) so I'd expect it to be slower.

Currently, [LLVM optimizes the to the same code](https://play.rust-lang.org/?gist=0e2832eb6863164940e77c71825d2b8c&version=nightly&mode=release&edition=2015) to the extend that they get deduplicated.^^